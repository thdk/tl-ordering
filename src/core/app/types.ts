import { Action } from "redux";
import { IOrderItemState } from "../orderitems/types";
import { IOrderState } from "../orders/types";
import { IProductState } from "../products/types";

export interface IState {
    orders: IOrderState;
    orderItems: IOrderItemState;
    products: IProductState;
}

export interface IAction<T, P = undefined> extends Action<T> {
    payload?: P;
}
