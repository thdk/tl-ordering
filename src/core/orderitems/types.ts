export interface IOrderItemDictionary { [orderId: string]: IOrderItem[]; }

export interface IOrderItemState { byOrderId: IOrderItemDictionary; }

export interface IOrderItem {
    readonly productId: string;
    readonly quantity: number;
}

// Types for deserialized api data

export interface IOrderItemData {
    productId: string;
    quantity: number;
    orderId: string;
}

// Types representing data from/for API

interface IApiData {
    [manyProps: string]: any;
}

// "product-id": "B102",
// "quantity": "5",
// "unit-price": "4.99",
// "total": "24.95"
export interface IApiOrderItem extends IApiData {
    quantity: string;
    total: string;
}
