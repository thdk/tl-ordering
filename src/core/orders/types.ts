import { IApiOrderItem, IOrderItemData } from "../orderitems/types";

export interface IOrderState {
    byId: IOrderDictionary;
    visibleIds: string[];
    isLoading: boolean;
}

export interface IOrderDictionary { [orderId: string]: IOrder; }

export interface IOrder {
    readonly id: string;
    readonly customerId: string;
    readonly isPlaced?: boolean;
}

// Types for deserialized api data

export interface IOrderData {
    id: string;
    items: IOrderItemData[];
    customerId: string;
}

export interface IPlaceOrderData {
    success: boolean;
    order: IOrderData;
    reason?: string;
}

// Types representing data from/for API

interface IApiData {
    [manyProps: string]: any;
}

// {
//  "id": "2",
//   "customer-id": "2",
//   "items": [
//     {
//       "product-id": "B102",
//       "quantity": "5",
//       "unit-price": "4.99",
//       "total": "24.95"
//     }
//   ],
//   "total": "24.95"
// }
export interface IApiOrder extends IApiData {
    id: string;
    items: IApiOrderItem[];
    total: string;
}

export interface IApiPlaceOrderResult extends IApiData {
    result: string;
    order: IApiOrder;
    reason?: string;
}
