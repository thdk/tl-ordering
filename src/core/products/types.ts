export interface IProductDictionary { [productId: string]: IProduct; }

export interface IProductState {
    byId: IProductDictionary;
    allIds: string[];
    isLoading: boolean;
}

export enum ProductCategory {
    "unknown" = 0,
    "tools" = 1,
    "electronics" = 2,
}

export interface IProduct {
    readonly id: string;
    readonly description: string;
    readonly category: ProductCategory;
    readonly price: number;
}

// Types for deserialized api data

export interface IProductData {
    readonly id: string;
    readonly description: string;
    readonly category: ProductCategory;
    readonly price: number;
}

// Types representing data from/for API

export interface IApiProduct {
    readonly id: string;
    readonly description: string;
    readonly category: string;
    readonly price: string;
}
