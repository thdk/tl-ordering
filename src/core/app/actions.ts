import { IOrderAction } from "../core/orders/actions";
import { IOrderItemAction } from "./orderItems";

export type AnyAction = IOrderAction | IOrderItemAction;
