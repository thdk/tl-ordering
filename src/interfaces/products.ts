export interface IProduct {
    id: string;
    unitPrice: number;
}