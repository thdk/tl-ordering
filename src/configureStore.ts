import { applyMiddleware, createStore } from "redux";
import logger from "redux-logger";
import thunkMiddleware from "redux-thunk";

import rootReducer from "./core/app/reducer";
import { IState } from "./core/app/types";

export function configureStore(initialState: Partial<IState>) {
    return createStore(
        rootReducer,
        initialState,
        applyMiddleware(
            thunkMiddleware,
            logger, // TODO: only add logger middleware in development
        ),
    );
}
