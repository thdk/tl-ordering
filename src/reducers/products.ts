import { IProductState } from "../interfaces/state";

export function products(state: IProductState = { byId: {}, allIds: [] }) {
    return state;
}