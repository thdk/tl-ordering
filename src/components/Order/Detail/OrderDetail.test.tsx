import React from "react";
import { Provider } from "react-redux";
import { BrowserRouter as Router } from "react-router-dom";
import renderer from "react-test-renderer";
import configureStore, { MockStoreEnhanced } from "redux-mock-store";

import { IState } from "../../../core/app/types";
import { ProductCategory } from "../../../core/products/types";
import ConnectedOrderDetail from "./OrderDetail";

const mockStore = configureStore<IState>();

const initialState: Pick<IState, "orderItems" | "orders" | "products"> = {
   orderItems: {
      byOrderId: {
         o1: [
            {
               productId: "p1",
               quantity: 3,
            },
            {
               productId: "p2",
               quantity: 2,
            },
         ],
         o2: [
            {
               productId: "p1",
               quantity: 7,
            },
         ],
         o3: [],
      },
   },
   orders: {
      byId: {
         o1: {
            id: "o1",
            customerId: "c1",
         },
         o2: {
            id: "o2",
            customerId: "c1",
         },
         o3: {
            id: "o3",
            customerId: "c1",
         },
      },
      visibleIds: ["o1", "o2", "o3"],
      isLoading: false,
   },
   products: {
      byId: {
         p1: {
            price: 10.99,
            description: "some product",
            category: ProductCategory.tools,
            id: "p1",
         },
      },
      allIds: ["p1"],
      isLoading: false,
   },
};

describe("OrderDetail rendering for a given state from redux store:", () => {
   let store: MockStoreEnhanced<IState>;
   beforeEach(() => {
      store = mockStore(initialState);
      store.dispatch = jest.fn();
   });

   describe("An order with multiple order items", () => {
      let component: renderer.ReactTestRenderer;

      beforeEach(() => {
         component = renderer.create(
            <Provider store={store}>
               <Router>
                  <ConnectedOrderDetail orderId="o2" />
               </Router>
            </Provider>,
         );
      });

      it("should match snapshot", () => {
         expect(component.toJSON()).toMatchSnapshot();
      });

      it("should display the total price for an order", () => {
         expect(
            component.root.findByProps({ className: "order-detail-total-value" }).children,
         ).toEqual(["10.99"]);
      });
   });

   describe("An order with zero order items", () => {
      let component: renderer.ReactTestRenderer;

      beforeEach(() => {
         component = renderer.create(
            <Provider store={store}>
               <Router>
                  <ConnectedOrderDetail orderId="o3" />
               </Router>
            </Provider>);
      });

      it("should match snapshot", () => {
         expect(component.toJSON()).toMatchSnapshot();
      });

      it("should display div with class order-detail--empty", () => {
         expect(component.root.findByProps({className: "order-detail--empty"})).toBeTruthy();
      });
   });

   describe("An order with one orderitem for an invalid product", () => {
      let component: renderer.ReactTestRenderer;

      beforeEach(() => {
         component = renderer.create(
            <Provider store={store}>
               <Router>
                  <ConnectedOrderDetail orderId="o1" />
               </Router>
            </Provider>);
      });

      it("should match snapshot", () => {
         expect(component.toJSON()).toMatchSnapshot();
      });

      it("should display the total price for an order", () => {
         expect(
            component.root.findByProps({ className: "order-detail-total-value" }).children,
         ).toEqual(["10.99"]);
      });
   });
});
