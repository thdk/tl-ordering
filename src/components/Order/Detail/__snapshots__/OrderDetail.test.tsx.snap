// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderDetail rendering for a given state from redux store: An order with multiple order items should match snapshot 1`] = `
Array [
  <div
    className="order-detail"
  >
    Order detail page for order 
    <span
      style={
        Object {
          "fontWeight": "bold",
        }
      }
    >
      o2
    </span>
    <hr />
    <br />
    Customer: 
    c1
  </div>,
  <ul>
    <li>
      <span
        className="orderitem-product-description"
      >
        some product
      </span>
      :
      <span
        className="orderitem-quantity"
      >
        7
      </span>
       x $
      10.99
       = $
      76.93
      <input
        onClick={[Function]}
        type="button"
        value="x"
      />
    </li>
  </ul>,
  <div
    className="order-detail-total"
  >
    <span
      className="order-detail-total-label"
    >
      Total:
    </span>
    <span
      className="order-detail-total-value"
    >
      10.99
    </span>
  </div>,
  <br />,
  <div>
    Add more products to this order:
    <form
      onSubmit={[Function]}
    >
      <label>
        Product id:
      </label>
      <select>
        $
        <option
          value="p1"
        >
          p1
        </option>
      </select>
      <label>
        Quantity:
      </label>
      <input
        defaultValue="1"
        min="0"
        type="number"
      />
      <input
        type="submit"
        value="Add product"
      />
    </form>
  </div>,
  <br />,
  <div>
    <input
      onClick={[Function]}
      type="button"
      value="Place order"
    />
    <br />
    Response printed in the developer console...
    <br />
    <a
      aria-current="page"
      className="active"
      href="/"
      onClick={[Function]}
      style={
        Object {
          "color": "black",
          "textDecoration": "none",
        }
      }
    >
      Back to order overview
    </a>
  </div>,
]
`;

exports[`OrderDetail rendering for a given state from redux store: An order with one orderitem for an invalid product should match snapshot 1`] = `
Array [
  <div
    className="order-detail"
  >
    Order detail page for order 
    <span
      style={
        Object {
          "fontWeight": "bold",
        }
      }
    >
      o1
    </span>
    <hr />
    <br />
    Customer: 
    c1
  </div>,
  <ul>
    <li>
      <span
        className="orderitem-product-description"
      >
        some product
      </span>
      :
      <span
        className="orderitem-quantity"
      >
        3
      </span>
       x $
      10.99
       = $
      32.97
      <input
        onClick={[Function]}
        type="button"
        value="x"
      />
    </li>
  </ul>,
  <div
    className="order-detail-total"
  >
    <span
      className="order-detail-total-label"
    >
      Total:
    </span>
    <span
      className="order-detail-total-value"
    >
      10.99
    </span>
  </div>,
  <br />,
  <div>
    Add more products to this order:
    <form
      onSubmit={[Function]}
    >
      <label>
        Product id:
      </label>
      <select>
        $
        <option
          value="p1"
        >
          p1
        </option>
      </select>
      <label>
        Quantity:
      </label>
      <input
        defaultValue="1"
        min="0"
        type="number"
      />
      <input
        type="submit"
        value="Add product"
      />
    </form>
  </div>,
  <br />,
  <div>
    <input
      onClick={[Function]}
      type="button"
      value="Place order"
    />
    <br />
    Response printed in the developer console...
    <br />
    <a
      aria-current="page"
      className="active"
      href="/"
      onClick={[Function]}
      style={
        Object {
          "color": "black",
          "textDecoration": "none",
        }
      }
    >
      Back to order overview
    </a>
  </div>,
]
`;

exports[`OrderDetail rendering for a given state from redux store: An order with zero order items should match snapshot 1`] = `
Array [
  <div
    className="order-detail"
  >
    Order detail page for order 
    <span
      style={
        Object {
          "fontWeight": "bold",
        }
      }
    >
      o3
    </span>
    <hr />
    <br />
    Customer: 
    c1
  </div>,
  <div
    className="order-detail--empty"
  >
    This order contains no order items
  </div>,
  <br />,
  <div>
    Add more products to this order:
    <form
      onSubmit={[Function]}
    >
      <label>
        Product id:
      </label>
      <select>
        $
        <option
          value="p1"
        >
          p1
        </option>
      </select>
      <label>
        Quantity:
      </label>
      <input
        defaultValue="1"
        min="0"
        type="number"
      />
      <input
        type="submit"
        value="Add product"
      />
    </form>
  </div>,
  <br />,
  <div>
    <input
      onClick={[Function]}
      type="button"
      value="Place order"
    />
    <br />
    Response printed in the developer console...
    <br />
    <a
      aria-current="page"
      className="active"
      href="/"
      onClick={[Function]}
      style={
        Object {
          "color": "black",
          "textDecoration": "none",
        }
      }
    >
      Back to order overview
    </a>
  </div>,
]
`;
